src/Middleware/RequestThrottling/test/TestUtils.cs

```diff
@@ -1,53 +1,96 @@
 // Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
 
+using System;
+using System.Threading;
 using System.Threading.Tasks;
 using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.RequestThrottling.Strategies;
 using Microsoft.Extensions.Logging.Abstractions;
 using Microsoft.Extensions.Options;
-using Microsoft.AspNetCore.RequestThrottling.Internal;
 
 namespace Microsoft.AspNetCore.RequestThrottling.Tests
 {
     public static class TestUtils
     {
-        public static RequestThrottlingMiddleware CreateTestMiddleware(int? maxConcurrentRequests, int requestQueueLimit = 5000, RequestDelegate onRejected = null, RequestDelegate next = null)
+        public static RequestThrottlingMiddleware CreateTestMiddleware(IQueuePolicy queue = null, RequestDelegate onRejected = null, RequestDelegate next = null)
         {
-            var options = new RequestThrottlingOptions
+            var options = Options.Create(new RequestThrottlingOptions
             {
-                MaxConcurrentRequests = maxConcurrentRequests,
-                RequestQueueLimit = requestQueueLimit
-            };
+                OnRejected = onRejected ?? (context => Task.CompletedTask),
+            });
+
+            return new RequestThrottlingMiddleware(
+                    next: next ?? (context => Task.CompletedTask),
+                    loggerFactory: NullLoggerFactory.Instance,
+                    queue: queue ?? CreateTailDropQueue(1, 0),
+                    options: options
+                );
+        }
 
-            return BuildFromOptions(options, onRejected, next);
+        public static RequestThrottlingMiddleware CreateTestMiddleware_TailDrop(int maxConcurrentRequests, int requestQueueLimit, RequestDelegate onRejected = null, RequestDelegate next = null)
+        {
+            return CreateTestMiddleware(
+                queue: CreateTailDropQueue(maxConcurrentRequests, requestQueueLimit),
+                onRejected: onRejected,
+                next: next
+                );
         }
 
-        public static RequestThrottlingMiddleware CreateBlockingTestMiddleware(int requestQueueLimit = 5000, RequestDelegate onRejected = null, RequestDelegate next = null)
+        internal static TailDrop CreateTailDropQueue(int maxConcurrentRequests, int requestQueueLimit = 5000)
         {
-            var options = new RequestThrottlingOptions
+            var options = Options.Create(new TailDropOptions
             {
-                MaxConcurrentRequests = 999,
-                RequestQueueLimit = requestQueueLimit,
-                ServerAlwaysBlocks = true
-            };
+                MaxConcurrentRequests = maxConcurrentRequests,
+                RequestQueueLimit = requestQueueLimit
+            });
 
-            return BuildFromOptions(options, onRejected, next);
+            return new TailDrop(options);
         }
+    }
+
+    public class TestStrategy : IQueuePolicy
```

Comment:
@anurse How do we feel about moq?

Reply:
I prefer not using it unless absolutely necessary :)

Reply:
This is a edgey case though. I prefer not using Moq because I prefer building "real fake" implementations of things (see `TestConnection` in SignalR, etc.) rather than just making enough things work so that the method doesn't throw. Here, the test strategy is basically just a way of pairing up two delegates into an `IQueuePolicy` so it's not much beyond Moq.

Unless we're using it elsewhere in this project, it still doesn't seem like this warrants adding a dependency.